File: apps/ui/src/lib/api/session-tasks.ts
// Session task registry API functions (knowledge/runtime-resources/session-tasks.md)

import { api } from "./client";
import type {
  PostTaskMessageRequest,
  SessionTask,
  SessionTaskDetail,
  SessionTaskState,
  TaskMessage,
} from "./types";

export interface ListSessionTasksOptions {
  state?: SessionTaskState;
  kind?: string;
}

/** List tasks owned by a session, optionally filtered by state/kind. */
export async function listSessionTasks(
  sessionId: string,
  options?: ListSessionTasksOptions,
): Promise<SessionTask[]> {
  const params = new URLSearchParams();
  if (options?.state) params.set("state", options.state);
  if (options?.kind) params.set("kind", options.kind);
  const queryString = params.toString();
  const response = await api.get<SessionTask[]>(
    `/v1/sessions/${sessionId}/tasks${queryString ? `?${queryString}` : ""}`,
  );
  return response.data;
}

export interface ListOrgTasksOptions {
  state?: SessionTaskState;
  kind?: string;
  /** Only tasks whose owning session's delegation-tree root is this session. */
  root_session_id?: string;
  /** RFC3339 timestamp; only tasks created at or after this time. */
  created_after?: string;
  /** Max tasks to return (server default 100, max 500, newest first). */
  limit?: number;
}

/** List background tasks across every session in the caller's org.
 *
 *  Backs the cross-session Work view (EVE-756). The server returns a flat,
 *  newest-first array; grouping into a delegation tree by `root_session_id` is a
 *  UI concern. */
export async function listOrgTasks(options?: ListOrgTasksOptions): Promise<SessionTask[]> {
  const params = new URLSearchParams();
  if (options?.state) params.set("state", options.state);
  if (options?.kind) params.set("kind", options.kind);
  if (options?.root_session_id) params.set("root_session_id", options.root_session_id);
  if (options?.created_after) params.set("created_after", options.created_after);
  if (options?.limit != null) params.set("limit", String(options.limit));
  const queryString = params.toString();
  const response = await api.get<SessionTask[]>(`/v1/tasks${queryString ? `?${queryString}` : ""}`);
  return response.data;
}

/** Get one task snapshot plus its recent message thread. */
export async function getSessionTask(
  sessionId: string,
  taskId: string,
): Promise<SessionTaskDetail> {
  const response = await api.get<SessionTaskDetail>(`/v1/sessions/${sessionId}/tasks/${taskId}`);
  return response.data;
}

/** Post an inbound message (steering or input answer) to a task.
 *  Not supported for subagent tasks; the server returns 400 for that kind. */
export async function postSessionTaskMessage(
  sessionId: string,
  taskId: string,
  request: PostTaskMessageRequest,
): Promise<TaskMessage> {
  const response = await api.post<TaskMessage>(
    `/v1/sessions/${sessionId}/tasks/${taskId}/messages`,
    request,
  );
  return response.data;
}

/** Record cooperative cancel intent. Returns the task snapshot. */
export async function cancelSessionTask(sessionId: string, taskId: string): Promise<SessionTask> {
  const response = await api.post<SessionTask>(`/v1/sessions/${sessionId}/tasks/${taskId}/cancel`);
  return response.data;
}

File: crates/server/src/api/session_tasks.rs
// Session task routes.
//
// Exposes the session task registry — background work owned by a session
// (subagents, external agent runs, background tools). See
// specs/session-tasks.md.

use crate::auth::{AuthState, ResolvedOrg};
use crate::domains::common::{Command, Ctx};
use crate::domains::session_tasks::{
    CancelSessionTask, GetSessionTask, ListSessionTasks, PostSessionTaskMessage, SessionTaskDetail,
};
use crate::services::EventService;
use crate::storage::StorageBackend;
use axum::{
    Json, Router,
    extract::{Path, Query, State},
    routing::{get, post},
};
use everruns_core::session_task::{TaskMessage, TaskMessagePart};
use everruns_core::{Caller, SessionTask};
use serde::Deserialize;
use std::sync::Arc;
use utoipa::ToSchema;

use super::common::{ApiResult, impl_auth_state};

/// App state for session task routes.
#[derive(Clone)]
pub struct AppState {
    pub db: Arc<StorageBackend>,
    pub auth: AuthState,
    pub event_service: Arc<EventService>,
}

impl AppState {
    pub fn new(db: Arc<StorageBackend>, auth: AuthState, event_service: Arc<EventService>) -> Self {
        Self {
            db,
            auth,
            event_service,
        }
    }

    fn ctx(&self, org: &ResolvedOrg) -> Ctx {
        let mut ctx = Ctx::minimal(
            Caller::from(org),
            self.db.clone(),
            None,
            self.auth.permission_resolver.clone(),
        );
        ctx.event_service = Some(self.event_service.clone());
        ctx
    }
}

impl_auth_state!(AppState);

pub fn routes(state: AppState) -> Router {
    Router::new()
        .route("/v1/sessions/{session_id}/tasks", get(list_tasks))
        .route("/v1/sessions/{session_id}/tasks/{task_id}", get(get_task))
        .route(
            "/v1/sessions/{session_id}/tasks/{task_id}/messages",
            post(post_task_message),
        )
        .route(
            "/v1/sessions/{session_id}/tasks/{task_id}/cancel",
            post(cancel_task),
        )
        .with_state(state)
}

#[derive(Debug, Default, Deserialize)]
pub struct ListTasksQuery {
    pub state: Option<String>,
    pub kind: Option<String>,
}

/// List background tasks owned by a session.
#[utoipa::path(
    get,
    path = "/v1/sessions/{session_id}/tasks",
    params(
        ("session_id" = String, Path, description = "Session ID"),
        ("state" = Option<String>, Query, description = "Filter by state"),
        ("kind" = Option<String>, Query, description = "Filter by kind"),
    ),
    responses(
        (status = 200, description = "Session tasks", body = Vec<SessionTask>),
        (status = 404, description = "Session not found"),
    ),
    tag = "session-tasks"
)]
pub async fn list_tasks(
    org: ResolvedOrg,
    State(state): State<AppState>,
    Path(session_id): Path<String>,
    Query(query): Query<ListTasksQuery>,
) -> ApiResult<Vec<SessionTask>> {
    Ok(Json(
        ListSessionTasks {
            session_id,
            state: query.state,
            kind: query.kind,
        }
        .run(&state.ctx(&org))
        .await?,
    ))
}

#[derive(Debug, Default, Deserialize)]
pub struct GetTaskQuery {
    pub after_id: Option<String>,
    pub limit: Option<u32>,
}

/// Get one session task with its recent message thread.
#[utoipa::path(
    get,
    path = "/v1/sessions/{session_id}/tasks/{task_id}",
    params(
        ("session_id" = String, Path, description = "Session ID"),
        ("task_id" = String, Path, description = "Task ID"),
        ("after_id" = Option<String>, Query, description = "Return only messages after this message ID (exclusive cursor)"),
        ("limit" = Option<u32>, Query, description = "Max messages to return (default 50, max 500)"),
    ),
    responses(
        (status = 200, description = "Session task detail", body = SessionTaskDetail),
        (status = 404, description = "Session or task not found"),
    ),
    tag = "session-tasks"
)]
pub async fn get_task(
    org: ResolvedOrg,
    State(state): State<AppState>,
    Path((session_id, task_id)): Path<(String, String)>,
    Query(query): Query<GetTaskQuery>,
) -> ApiResult<SessionTaskDetail> {
    Ok(Json(
        GetSessionTask {
            session_id,
            task_id,
            after_id: query.after_id,
            limit: query.limit,
        }
        .run(&state.ctx(&org))
        .await?,
    ))
}

/// Request body for posting an inbound task message.
#[derive(Debug, Deserialize, ToSchema)]
pub struct PostTaskMessageBody {
    /// Plain-text message (alternative to `content`).
    #[serde(default)]
    pub text: Option<String>,
    /// Structured message parts (alternative to `text`).
    #[serde(default)]
    pub content: Option<Vec<TaskMessagePart>>,
    /// Input request ID this message answers, when applicable.
    #[serde(default)]
    pub in_reply_to: Option<String>,
}

/// Send an inbound message to a session task.
#[utoipa::path(
    post,
    path = "/v1/sessions/{session_id}/tasks/{task_id}/messages",
    params(
        ("session_id" = String, Path, description = "Session ID"),
        ("task_id" = String, Path, description = "Task ID"),
    ),
    request_body = PostTaskMessageBody,
    responses(
        (status = 200, description = "Recorded task message", body = TaskMessage),
        (status = 400, description = "Empty message"),
        (status = 404, description = "Session or task not found"),
    ),
    tag = "session-tasks"
)]
pub async fn post_task_message(
    org: ResolvedOrg,
    State(state): State<AppState>,
    Path((session_id, task_id)): Path<(String, String)>,
    Json(body): Json<PostTaskMessageBody>,
) -> ApiResult<TaskMessage> {
    Ok(Json(
        PostSessionTaskMessage {
            session_id,
            task_id,
            text: body.text,
            content: body.content,
            in_reply_to: body.in_reply_to,
        }
        .run(&state.ctx(&org))
        .await?,
    ))
}

/// Request cooperative cancellation of a session task.
#[utoipa::path(
    post,
    path = "/v1/sessions/{session_id}/tasks/{task_id}/cancel",
    params(
        ("session_id" = String, Path, description = "Session ID"),
        ("task_id" = String, Path, description = "Task ID"),
    ),
    responses(
        (status = 200, description = "Task with cancel intent recorded", body = SessionTask),
        (status = 404, description = "Session or task not found"),
    ),
    tag = "session-tasks"
)]
pub async fn cancel_task(
    org: ResolvedOrg,
    State(state): State<AppState>,
    Path((session_id, task_id)): Path<(String, String)>,
) -> ApiResult<SessionTask> {
    Ok(Json(
        CancelSessionTask {
            session_id,
            task_id,
        }
        .run(&state.ctx(&org))
        .await?,
    ))
}
